feat(pdk) add documents for JavaScript and Python PDK (2.4) (#2690)
Kong/docs.konghq.com: app/gateway-oss/2.3.x/external-plugins.md

pluginserver_other_one_query_cmd = /usr/local/bin/other-one -dump
```

## Developing JavaScript plugins

{{site.base_gateway}} support for the JavaScript language is provided by [kong-js-pdk].
The library provides a plugin server to provide runtime for JavaScript plugins, and
functions to access {{site.base_gateway}} features of the [PDK][kong-pdk].

TypeScript is also supported in the following ways:

> - [kong-js-pdk] includes type definitions for PDK functions that allows type checking

Reviewer: ```suggestion
- [kong-js-pdk] includes type definitions for PDK functions that allow type checking
```

when developing plugins in TypeScript.

> - Plugin written in TypeScript can be loaded directly and transpiled on-the-fly.

Reviewer: ```suggestion
- Plugin written in TypeScript can be loaded directly and transpiled on the fly.
```

### Development

> [kong-js-pdk] can be installed via `npm`. To install the plugin server binary globally, use:

Reviewer: ```suggestion
[kong-js-pdk] can be installed using `npm`. To install the plugin server binary globally:
```

```
npm install kong-pdk -g
```

> To use it in your local development directory, use:

Reviewer: ```suggestion
To install it in your local development directory:
```

```
npm install kong-pdk --save
```

> A valid JavaScript plugin implementation should export following object:

Reviewer: ```suggestion
A valid JavaScript plugin implementation should export the following object:
```

schema of plugin, it shares the same syntax as it's a Lua plugin. `Version` and `Priority`
defines the version number and priority of execution respectively.

**Note**: Check out [this repository](https://github.com/Kong/kong-js-pdk/tree/master/examples)
for example JavaScript and TypeScript plugins.

#### Phase Handlers

Similarly to {{site.base_gateway}} Lua plugins, you can implement custom logic to be executed at
various points of the request processing lifecycle. For example, to execute
custom Javascript code in the access phase, define a function named `access`:

```javascript
class KongPlugin {
  constructor(config) {
    this.config = config
  }

  async access(kong) {
    // ...
  }
}
```

The phases you can implement custom logic for are as follows, and the expected function
signature is the same for all of them:
- `certificate`
- `rewrite`
- `access`
- `response`
- `preread`
- `log`

Similar to Lua plugins, the presence of the `response` handler automatically enables the buffered proxy mode.

#### PDK functions

[kong-js-pdk] invokes PDK functions in Kong through network-based RPC. Thus each function returns a Promise

> instance; it's convient to use `async`/`await` keywords in phase handlers for better readibility.

Reviewer: ```suggestion
instance; it's convenient to use `async`/`await` keywords in phase handlers for better readability.
```
